libraries/AP_PacketDigitalCAN/AP_PacketDigitalCAN.cpp

```diff
@@ -0,0 +1,609 @@
+    }
+}
+
+// get MPPT_device index by serial number.
+// if serial number found in MPP_devices list, return the 0 indexed value
+// else return UINT8_MAX
+uint8_t AP_PacketDigitalCAN::get_device_index(const uint16_t serial_number) const
+{
+    for (uint8_t i=0; i<device_count; i++) {
+        if (MPPT_devices[i].serialnumber == serial_number) {
+            return i;
+        }
+    }
+    return UINT8_MAX;
+}
+
+// write frame on CAN bus, returns true on success
+bool AP_PacketDigitalCAN::write_frame(AP_HAL::CANFrame &out_frame, const uint64_t timeout)
+{
+    if (!_initialized) {
+        debug_can(AP_CANManager::LOG_ERROR, "Driver not initialized for write_frame\n\r");
+        return false;
+    }
+
+    bool read_select = false;
+    bool write_select = true;
+    bool ret;
+    do {
+        ret = _can_iface->select(read_select, write_select, &out_frame, AP_HAL::native_micros64() + timeout);
+        if (!ret || !write_select) {
+            hal.scheduler->delay_microseconds(200);
+        }
+    } while (!ret || !write_select);
+
+    uint64_t deadline = AP_HAL::native_micros64() + 2000000;
+    return (_can_iface->send(out_frame, deadline, AP_HAL::CANIface::AbortOnError) == 1);
+}
+
+// read frame on CAN bus, returns true on succses
+bool AP_PacketDigitalCAN::read_frame(AP_HAL::CANFrame &recv_frame, const uint64_t timeout, AP_HAL::CANIface::CanIOFlags &flags)
+{
+    if (!_initialized) {
+        debug_can(AP_CANManager::LOG_ERROR, "Driver not initialized for read_frame\n\r");
+        return false;
+    }
+    // wait for space in buffer to read
+    bool read_select = true;
+    bool write_select = false;
+    bool ret = _can_iface->select(read_select, write_select, nullptr, AP_HAL::native_micros64() + timeout);
+    if (!ret || !read_select) {
+        // return false if no data is available to read
+        return false;
+    }
+
+    uint64_t time;
+
+    // read frame and return success
+    return (_can_iface->receive(recv_frame, time, flags) == 1);
+}
+
+bool AP_PacketDigitalCAN::pre_arm_check(char* reason, uint8_t reason_len) const
+{
+    for (uint8_t i=0; i<device_count; i++) {
+        if (!is_healthy_by_index(i)) {
+            return false;
+        }
+    }
+    return true;
+}
+
+
+// return fault code as string
+const char* AP_PacketDigitalCAN::get_fault_code_string(const Fault_Flags fault) const
+{
+    switch (fault) {
+    case Fault_Flags::NONE:
+        return "None";
+    case Fault_Flags::OVER_VOLTAGE:
+        return "Over-Voltage";
+    case Fault_Flags::UNDER_VOLTAGE:
+        return "Under-Voltage";
+    case Fault_Flags::OVER_CURRENT:
+        return "Over-Current";
+    case Fault_Flags::OVER_TEMPERATURE:
+        return "Over-Temperature";
+    default:
+        return "Unknown";
+    }
+}
+
+bool AP_PacketDigitalCAN::get_voltage_and_current_and_temp(const uint16_t serialnumber, const bool chkInput, float &voltage, float &current, float &temperature) const
```

Comment:
Some description of what the whole chkInput thing is all about would be good here.

Reply:
sure, added.